File: backend-dotnet/FinanceTracker.Api/Services/GoalService.cs
using FinanceTracker.Api.Contracts;
using FinanceTracker.Api.Data;
using FinanceTracker.Api.Entities;
using FinanceTracker.Api.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace FinanceTracker.Api.Services;

public sealed class GoalService(
    FinanceTrackerDbContext dbContext,
    CurrentUserService currentUserService,
    AccountService accountService,
    LedgerService ledgerService)
{
    public async Task<IReadOnlyList<GoalResponse>> GetAllAsync(CancellationToken cancellationToken)
    {
        var goals = await dbContext.Goals
            .AsNoTracking()
            .Where(goal => goal.UserId == currentUserService.GetCurrentUserId())
            .OrderBy(goal => goal.TargetDate)
            .ToListAsync(cancellationToken);

        return goals.Select(ToResponse).ToList();
    }

    public async Task<GoalResponse> CreateAsync(GoalRequest request, CancellationToken cancellationToken)
    {
        var goal = new Goal
        {
            UserId = currentUserService.GetCurrentUserId(),
            Name = request.name!,
            TargetAmount = request.targetAmount!.Value,
            CurrentAmount = 0m,
            TargetDate = request.targetDate,
            Status = request.status ?? GoalStatus.ACTIVE
        };

        dbContext.Goals.Add(goal);
        await dbContext.SaveChangesAsync(cancellationToken);
        return ToResponse(goal);
    }

    public async Task<GoalResponse> UpdateAsync(Guid id, GoalRequest request, CancellationToken cancellationToken)
    {
        var goal = await GetGoalAsync(id, currentUserService.GetCurrentUserId(), cancellationToken);
        goal.Name = request.name!;
        goal.TargetAmount = request.targetAmount!.Value;
        goal.TargetDate = request.targetDate;
        goal.Status = request.status ?? goal.Status;
        RecalculateStatus(goal);

        await dbContext.SaveChangesAsync(cancellationToken);
        return ToResponse(goal);
    }

    public async Task<GoalResponse> ContributeAsync(Guid id, GoalContributionRequest request, CancellationToken cancellationToken)
    {
        var userId = currentUserService.GetCurrentUserId();
        var goal = await GetGoalAsync(id, userId, cancellationToken);
        var account = await accountService.GetAccountAsync(request.accountId!.Value, userId, cancellationToken);

        if (account.CurrentBalance < request.amount!.Value)
        {
            throw new BadRequestException("Insufficient account balance");
        }

        var user = await dbContext.Users.FirstAsync(candidate => candidate.Id == userId, cancellationToken);
        goal.CurrentAmount += request.amount.Value;
        RecalculateStatus(goal);

        await using var transaction = await dbContext.Database.BeginTransactionAsync(cancellationToken);
        ledgerService.CreateTransaction(
            user,
            account,
            null,
            TransactionType.TRANSFER_OUT,
            request.amount.Value,
            DateOnly.FromDateTime(DateTime.UtcNow),
            goal.Name,
            $"Contribution to goal: {goal.Name}",
            "GOAL",
            Guid.NewGuid()
        );

        await dbContext.SaveChangesAsync(cancellationToken);
        await transaction.CommitAsync(cancellationToken);
        return ToResponse(goal);
    }

    public async Task<GoalResponse> WithdrawAsync(Guid id, GoalContributionRequest request, CancellationToken cancellationToken)
    {
        var userId = currentUserService.GetCurrentUserId();
        var goal = await GetGoalAsync(id, userId, cancellationToken);
        if (goal.CurrentAmount < request.amount!.Value)
        {
            throw new BadRequestException("Insufficient goal balance");
        }

        var account = await accountService.GetAccountAsync(request.accountId!.Value, userId, cancellationToken);
        var user = await dbContext.Users.FirstAsync(candidate => candidate.Id == userId, cancellationToken);

        goal.CurrentAmount -= request.amount.Value;
        RecalculateStatus(goal);

        await using var transaction = await dbContext.Database.BeginTransactionAsync(cancellationToken);
        ledgerService.CreateTransaction(
            user,
            account,
            null,
            TransactionType.TRANSFER_IN,
            request.amount.Value,
            DateOnly.FromDateTime(DateTime.UtcNow),
            goal.Name,
            $"Withdrawal from goal: {goal.Name}",
            "GOAL",
            Guid.NewGuid()
        );

        await dbContext.SaveChangesAsync(cancellationToken);
        await transaction.CommitAsync(cancellationToken);
        return ToResponse(goal);
    }

    public async Task<Goal> GetGoalAsync(Guid id, Guid userId, CancellationToken cancellationToken)
    {
        return await dbContext.Goals.FirstOrDefaultAsync(goal => goal.Id == id && goal.UserId == userId, cancellationToken)
               ?? throw new NotFoundException("Goal not found");
    }

    private static void RecalculateStatus(Goal goal)
    {
        if (goal.CurrentAmount >= goal.TargetAmount)
        {
            goal.Status = GoalStatus.COMPLETED;
        }
        else if (goal.Status == GoalStatus.COMPLETED)
        {
            goal.Status = GoalStatus.ACTIVE;
        }
    }

    private static GoalResponse ToResponse(Goal goal)
    {
        var progressPercent = goal.TargetAmount == 0
            ? 0m
            : Math.Round(goal.CurrentAmount * 100m / goal.TargetAmount, 2, MidpointRounding.AwayFromZero);

        return new GoalResponse(
            goal.Id,
            goal.Name,
            goal.TargetAmount,
            goal.CurrentAmount,
            goal.TargetDate,
            goal.Status,
            progressPercent
        );
    }
}

File: backend/src/main/java/com/finance/services/GoalService.java
package com.finance.services;

import com.finance.dto.GoalDtos.GoalContributionRequest;
import com.finance.dto.GoalDtos.GoalRequest;
import com.finance.dto.GoalDtos.GoalResponse;
import com.finance.entities.Goal;
import com.finance.entities.GoalStatus;
import com.finance.entities.TransactionType;
import com.finance.exception.BadRequestException;
import com.finance.exception.NotFoundException;
import com.finance.repositories.GoalRepository;
import com.finance.repositories.UserRepository;
import com.finance.security.CurrentUserService;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class GoalService {

    private final GoalRepository goalRepository;
    private final UserRepository userRepository;
    private final CurrentUserService currentUserService;
    private final AccountService accountService;
    private final LedgerService ledgerService;

    @Transactional(readOnly = true)
    public List<GoalResponse> getAll() {
        return goalRepository.findByUserIdOrderByTargetDateAsc(currentUserService.getCurrentUserId())
                .stream()
                .map(this::toResponse)
                .toList();
    }

    @Transactional
    public GoalResponse create(GoalRequest request) {
        Goal goal = new Goal();
        goal.setUser(userRepository.findById(currentUserService.getCurrentUserId()).orElseThrow(() -> new NotFoundException("User not found")));
        goal.setName(request.name());
        goal.setTargetAmount(request.targetAmount());
        goal.setCurrentAmount(BigDecimal.ZERO);
        goal.setTargetDate(request.targetDate());
        goal.setStatus(request.status() == null ? GoalStatus.ACTIVE : request.status());
        return toResponse(goalRepository.save(goal));
    }

    @Transactional
    public GoalResponse update(UUID id, GoalRequest request) {
        Goal goal = getGoal(id, currentUserService.getCurrentUserId());
        goal.setName(request.name());
        goal.setTargetAmount(request.targetAmount());
        goal.setTargetDate(request.targetDate());
        goal.setStatus(request.status() == null ? goal.getStatus() : request.status());
        recalculateStatus(goal);
        return toResponse(goalRepository.save(goal));
    }

    @Transactional
    public GoalResponse contribute(UUID id, GoalContributionRequest request) {
        UUID userId = currentUserService.getCurrentUserId();
        Goal goal = getGoal(id, userId);
        var account = accountService.getAccount(request.accountId(), userId);
        if (account.getCurrentBalance().compareTo(request.amount()) < 0) {
            throw new BadRequestException("Insufficient account balance");
        }

        goal.setCurrentAmount(goal.getCurrentAmount().add(request.amount()));
        recalculateStatus(goal);
        ledgerService.createTransaction(
                goal.getUser(),
                account,
                null,
                TransactionType.TRANSFER_OUT,
                request.amount(),
                LocalDate.now(),
                goal.getName(),
                "Contribution to goal: " + goal.getName(),
                "GOAL",
                UUID.randomUUID()
        );
        return toResponse(goalRepository.save(goal));
    }

    @Transactional
    public GoalResponse withdraw(UUID id, GoalContributionRequest request) {
        UUID userId = currentUserService.getCurrentUserId();
        Goal goal = getGoal(id, userId);
        if (goal.getCurrentAmount().compareTo(request.amount()) < 0) {
            throw new BadRequestException("Insufficient goal balance");
        }
        var account = accountService.getAccount(request.accountId(), userId);
        goal.setCurrentAmount(goal.getCurrentAmount().subtract(request.amount()));
        recalculateStatus(goal);
        ledgerService.createTransaction(
                goal.getUser(),
                account,
                null,
                TransactionType.TRANSFER_IN,
                request.amount(),
                LocalDate.now(),
                goal.getName(),
                "Withdrawal from goal: " + goal.getName(),
                "GOAL",
                UUID.randomUUID()
        );
        return toResponse(goalRepository.save(goal));
    }

    @Transactional(readOnly = true)
    public Goal getGoal(UUID id, UUID userId) {
        return goalRepository.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new NotFoundException("Goal not found"));
    }

    private void recalculateStatus(Goal goal) {
        if (goal.getCurrentAmount().compareTo(goal.getTargetAmount()) >= 0) {
            goal.setStatus(GoalStatus.COMPLETED);
        } else if (goal.getStatus() == GoalStatus.COMPLETED) {
            goal.setStatus(GoalStatus.ACTIVE);
        }
    }

    private GoalResponse toResponse(Goal goal) {
        BigDecimal progress = goal.getTargetAmount().compareTo(BigDecimal.ZERO) == 0
                ? BigDecimal.ZERO
                : goal.getCurrentAmount().multiply(BigDecimal.valueOf(100))
                .divide(goal.getTargetAmount(), 2, RoundingMode.HALF_UP);
        return new GoalResponse(
                goal.getId(),
                goal.getName(),
                goal.getTargetAmount(),
                goal.getCurrentAmount(),
                goal.getTargetDate(),
                goal.getStatus(),
                progress
        );
    }
}
